blog/posts/dune-universal-libraries-preview.md

```diff
@@ -0,0 +1,241 @@
++ (only_sources)
+  (files {a,b}.ml{,i}))
+```
+
+
+## Checkpoint
+
+Our Dune file allows us to move forward. We were now able to define multiple
+libraries that share common implementations across native code and Melange.
+Though library names still need to be different. And, overall, we still face
+some other glaring limitations:
+
+- The `(wrapped false)` requirement makes it impossible to namespace these
+  libraries;
+- Defining libraries in different directories and using `copy_files` places
+  extra separation between common implementations, and adds extra build
+  configuration overhead;
+- Publishing a library with `(modes :standard melange)` adds a non-optional
+  dependency on Melange, which should really be optional for native-only
+  consumers.
+
+## Testing a New Solution
+
+We became intentful on removing these limitations, and realized at some point
+that our use case is somewhat similar to cross-compilation, which Dune [already
+supports well](https://dune.readthedocs.io/en/stable/cross-compilation.html).
+The key insight, which we shared in a Dune proposal
+([dune#10222](https://github.com/ocaml/dune/issues/10222)), is that we could
+share library names as long as they resolved to a single library per [build
+context](https://dune.readthedocs.io/en/stable/reference/dune-workspace/context.html).
+
+After making the proposed changes to Dune
+([dune#10220](https://github.com/ocaml/dune/pull/10220),
+[dune#10307](https://github.com/ocaml/dune/pull/10307),
+[dune#10354](https://github.com/ocaml/dune/pull/10354),
+[dune#10355](https://github.com/ocaml/dune/pull/10355)) we found ourselves
+having implemented support for:
+
+- Dune libraries with the same name;
+- which may be defined in the same directory;
+- as long as they don't conflict in the same context.
+    - to achieve that, we use e.g. `(enabled_if (= %{context_name} melange))`.
+
+Putting it all together, our example can be adapted to look like:
+
+```clj
+;; src/dune
+(library
+ (name a)
+ (modules a b c)
+ (enabled_if
+  (= %{context_name} default)))
+
+(library
+ (name a)
+ (modes melange)
+ (modules a b c)
+ (enabled_if
+  (= %{context_name} melange)))
+```
+
+In other words, we define two libraries named `a`, each in their own build
+context (with build artifacts ending up in `_build/default` and
+`_build/melange`). In the `melange` context, the library has `(modes melange)`.
+
+Both libraries contain modules `A`, `B` and `C` like before. Their
+corresponding source files can live in a single directory, no copying required.
+If we need to vary `C`'s implementation, we can express that in Dune rules:
+
+```clj
+(rule
+ (target c.ml)
+ (deps c.native.ml)
+ (action
+  (with-stdout-to
+   %{target}
+   (echo "let backend = \"OCaml\"")))
+ (enabled_if
+  (= %{context_name} default)))
+
+(rule
+ (target c.ml)
+ (deps c.melange.ml)
+ (action
+  (with-stdout-to
+   %{target}
+   (echo "let backend = \"Melange\"")))
+ (enabled_if
+  (= %{context_name} melange)))
```

Comment:
I'd rather keep it simple while we're under "preview" (I agree with your other comment that we should say this is a preview).

The goal of this post is to tell the story rather than prescribe a dune build setup; I'd love to follow up with a complete tutorial / part 2 once we get it right.

Reply:
how about I add a paragraph mentioning just that?

that it's an illustrative example that we don't know whether it's going to work

Reply:
Would be helpful 👍